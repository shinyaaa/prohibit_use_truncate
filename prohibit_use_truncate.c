#include "postgres.h"
#include "fmgr.h"
#include "tcop/utility.h"

PG_MODULE_MAGIC;

void _PG_init(void);
void _PG_fini(void);

/* Saved hook values in case of unload */
static ProcessUtility_hook_type prev_ProcessUtility = NULL;

static void
prohibit_use_truncate(PlannedStmt *pstmt, const char *queryString,
					ProcessUtilityContext context,
					ParamListInfo params, QueryEnvironment *queryEnv,
					DestReceiver *dest, char *completionTag)
{
	Node *parsetree = pstmt->utilityStmt;
	if (nodeTag(parsetree) == T_TruncateStmt)
		elog(ERROR, "Don't use TRUNCATE.");
	else
		standard_ProcessUtility(pstmt, queryString,
								context, params, queryEnv,
								dest, completionTag);
}

void
_PG_init(void)
{
	prev_ProcessUtility = ProcessUtility_hook;
	ProcessUtility_hook = prohibit_use_truncate;
}

void
_PG_fini(void)
{
	ProcessUtility_hook = prev_ProcessUtility;
}