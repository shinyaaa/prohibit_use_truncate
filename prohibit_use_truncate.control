# prohibit_use_truncate extension
comment = 'prohibit_use_truncate is add and submodule'
default_version = '1.0'
module_pathname = '$libdir/prohibit_use_truncate'
relocatable = true