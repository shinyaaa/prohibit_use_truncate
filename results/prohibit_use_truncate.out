CREATE EXTENSION prohibit_use_truncate;
ERROR:  syntax error at or near "AS"
LINE 1: CREATE EXTENSION prohibit_use_truncate;
                                            ^
CREATE TABLE test_table (id integer, number integer);
INSERT INTO test_table VALUES (1, 10);
TRUNCATE test_table;
SELECT * FROM test_table;
 id | number 
----+--------
(0 rows)

DROP TABLE test_table;
DROP EXTENSION prohibit_use_truncate;
ERROR:  extension "prohibit_use_truncate" does not exist
