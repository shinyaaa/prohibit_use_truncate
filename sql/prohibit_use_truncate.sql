CREATE EXTENSION prohibit_use_truncate;
CREATE TABLE test_table (id integer, number integer);
INSERT INTO test_table VALUES (1, 10);
TRUNCATE test_table;
SELECT * FROM test_table;
DROP TABLE test_table;
DROP EXTENSION prohibit_use_truncate;