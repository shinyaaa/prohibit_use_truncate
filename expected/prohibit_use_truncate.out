CREATE EXTENSION prohibit_use_truncate;
CREATE TABLE test_table (id integer, number integer);
INSERT INTO test_table VALUES (1, 10);
TRUNCATE test_table;
ERROR:  Don't use TRUNCATE.
SELECT * FROM test_table;
 id | number 
----+--------
  1 |     10
(1 rows)

DROP TABLE test_table;
DROP EXTENSION prohibit_use_truncate;